Framework/PythonInterface/plugins/algorithms/SaveNexusPD.py

```diff
@@ -0,0 +1,309 @@
+            elif self._sourcePos is not None:
+                wsname = '__SaveNexusPD_%s' % target
+                self._tempNames.append(wsname)
+                temp = api.ConvertUnits(InputWorkspace=wksp,
+                                        OutputWorkspace=wsname,
+                                        Target=target, EMode='Elastic')
+                result[i] = temp
+            else:
+                pass   # can't ConvertUnits
+        return result
+
+    def _determineSourceSample(self, wksp):
+        self._sample = wksp.getInstrument().getSample()
+        source = wksp.getInstrument().getSource()
+
+        # set all of the information to None if the
+        # instrument doesn't supply it
+        if (source is not None) and (self._sample is not None):
+            self._sourcePos = self._sample.getPos() - source.getPos()
+
+    #pylint: disable=too-many-branches
+    def PyExec(self):
+        self._determineDtype()
+        self._determineCompression()
+
+        wksp = self.getProperty('InputWorkspace').value
+
+        self._determineSourceSample(wksp)
+
+        (tof, dspacing, momentumtransfer) = self._getOtherUnits(wksp)
+
+        xAxesToWrite = []
+        if tof is not None:
+            xAxesToWrite.append('tof')
+        if dspacing is not None:
+            xAxesToWrite.append('dspacing')
+        if self.getProperty('WriteMomentumTransfer').value and \
+           momentumtransfer is not None:
+            xAxesToWrite.append('Q')
+
+        append = self.getProperty('Append').value
+        if append:
+            filemode = 'a'  # append
+        else:
+            filemode = 'w'  # write, will silently overwrite
+
+        filename = self.getProperty('OutputFilename').value
+        with h5py.File(filename, filemode) as handle:
+            # default name for entry is the workspace name
+            wkspname = self.getProperty('NXentry').value.strip()
+            if len(wkspname) == 0:
+                wkspname = str(wksp)
+
+            # check for the entry alread existing in append mode
+            if append and wkspname in handle.keys():
+                raise IOError("NXentry named '%s' already exists in '%s'" % \
+                      (wkspname, filename))
+
+            # create the entry
+            nxentry = handle.create_group(wkspname)
+            nxentry.attrs[self.NX_CLASS] = 'NXentry'
+
+            self._writeProtonCharge(nxentry, wksp)
+
+            nxinstrument = self._createInstrument(nxentry)
+
+            for i in xrange(wksp.getNumberHistograms()):
+                writeDx = not np.all(wksp.readDx(i) == 0)
+
+                dataname = "spectrum_%d" % wksp.getSpectrum(i).getSpectrumNo()
+
+                nxdata = nxentry.create_group(dataname)
+                nxdata.attrs[self.NX_CLASS] = 'NXdata'
+
+                if self._sourcePos is None:
+                    detector = None
+                else:
+                    detector = wksp.getDetector(i)
```

Comment:
I think this needs a `try..except RuntimeError:` wrapped around it, with `detector = None` for the case where `getDetector` throws. Some ISIS files have more data than has been put in the IDF, one example is `IRS38633`  in the `Data/UnitTest` directory.